File: go/algorithms/stooge.go
package algorithms

// StoogeSort sorts in-place recursively.
func StoogeSort(arr []int) {
    stoogeSortRec(arr, 0, len(arr)-1)
}

func stoogeSortRec(arr []int, l, h int) {
    if l >= h {
        return
    }
    if arr[l] > arr[h] {
        arr[l], arr[h] = arr[h], arr[l]
    }
    if h-l+1 > 2 {
        t := (h - l + 1) / 3
        stoogeSortRec(arr, l, h-t)
        stoogeSortRec(arr, l+t, h)
        stoogeSortRec(arr, l, h-t)
    }
}

File: rust/src/stooge.rs
pub fn stooge_sort(arr: &mut [i32]) {
    if arr.is_empty() {
        return;
    }
    let len = arr.len();
    stooge_sort_rec(arr, 0, len - 1);
}

fn stooge_sort_rec(arr: &mut [i32], l: usize, h: usize) {
    if l >= h {
        return;
    }
    if arr[l] > arr[h] {
        arr.swap(l, h);
    }
    let len = h - l + 1;
    if len > 2 {
        let t = len / 3;
        stooge_sort_rec(arr, l, h - t);
        stooge_sort_rec(arr, l + t, h);
        stooge_sort_rec(arr, l, h - t);
    }
}
